SwiftCompilerSources/Sources/Optimizer/Utilities/EscapeInfoWalker.swift

```diff
@@ -0,0 +1,834 @@
+/// * Starting at allocations, walks down from defs to uses ("Where does the value go to?")
+/// * Starting at stores, walks up from uses to defs ("Were does the value come from?")
+///
+/// The result of the walk indicates if the initial value "escapes" or not.
+/// The value escapes if the walk reaches a point where the further flow of the
+/// value cannot be tracked anymore.
+/// Example:
+/// \code
+///   %1 = alloc_ref $X    // 1. initial value: walk down to the `store`
+///   %2 = alloc_stack $X  // 3. walk down to %3
+///   store %1 to %2       // 2. walk up to `%2`
+///   %3 = load %2         // 4. continue walking down to the `return`
+///   return %3            // 5. The value is escaping!
+/// \endcode
+///
+/// During the walk, a projection path indicates where the initial value is
+/// contained in an aggregate.
+/// Example for a walk-down:
+/// \code
+///   %1 = alloc_ref                   // 1. initial value, path = empty
+///   %2 = struct $S (%1)              // 2. path = s0
+///   %3 = tuple (%other, %1)          // 3. path = t1.s0
+///   %4 = tuple_extract %3, 1         // 4. path = s0
+///   %5 = struct_extract %4, #field   // 5. path = empty
+/// \endcode
+///
+/// The `followStores` flag, which is passed together with the path, indicates
+/// if stored values should be included in the walk.
+/// If the initial value is stored to some memory allocation, we usually don't
+/// care if other values are stored to that location as well. Example:
+/// \code
+///   %1 = alloc_ref $X    // 1. initial value, walk down to the `store`
+///   %2 = alloc_stack $X  // 3. walk down to the second `store`
+///   store %1 to %2       // 2. walk up to %2
+///   store %other to %2   // 4. ignore (followStores == false): %other doesn't impact the "escapeness" of %1
+/// \endcode
+///
+/// But once the the up-walk sees a load, it has to follow stores from that point on.
+/// Example:
+/// \code
+/// bb0(%function_arg):            // 7. escaping! %1 escapes through %function_arg
+///   %1 = alloc_ref $X            // 1. initial value, walk down to the second `store`
+///   %addr = alloc_stack %X       // 5. walk down to the first `store`
+///   store %function_arg to %addr // 6. walk up to %function_arg (followStores == true)
+///   %2 = load %addr              // 4. walk up to %addr, followStores = true
+///   %3 = ref_element_addr %2, #f // 3. walk up to %2
+///   store %1 to %3               // 2. walk up to %3
+/// \endcode
+///
+/// The algorithm doesn't distinguish between addresses and values, i.e. loads
+/// and stores are treated as simple forwarding instructions, like casts.
+/// For escaping it doesn't make a difference if a value or an address pointing to
+/// the value, escapes.
+/// An exception are `isEscaping(address: Value)` and similar functions: they ignore
+/// values which are loaded from the address in question.
+struct EscapeInfoWalker : ValueDefUseWalker, AddressDefUseWalker, ValueUseDefWalker, AddressUseDefWalker {
```

Comment:
As I said in a previous comment, it's not good that all the internal conforming functions need to be public.
Therefore I suggest to make a public struct, e.g. `EscapeInfo` which contains a private `EscapeInfoVisitor`, which conforms to DefVisitor and UseVisitor.
Then there is a clear separation between public API and internal implementation.